chemprop/hyperopt_utils.py

```diff
@@ -158,13 +220,21 @@ def load_manual_trials(manual_trials_dirs: List[str], param_keys: List[str], hyp
             trial_args = json.load(f)
 
         # Check for differences in manual trials and hyperopt space
-        if 'hidden_size' in param_keys:
+        if 'linked_hidden_size' in param_keys:
             if trial_args['hidden_size'] != trial_args['ffn_hidden_size']:
                 raise ValueError(f'The manual trial in {trial_dir} has a hidden_size {trial_args["hidden_size"]} '
                 f'that does not match its ffn_hidden_size {trial_args["ffn_hidden_size"]}, as it would in hyperparameter search.')
-        for arg in matching_args:
-            if arg not in param_keys:
-                if getattr(hyperopt_args,arg) != trial_args[arg]:
+        elif 'hidden_size' not in param_keys or 'ffn_hidden_size' not in param_keys:
+            if 'hidden_size' not in param_keys:
+                if getattr(hyperopt_args,'hidden_size') != trial_args['hidden_size']:
+                    raise ValueError(f'Manual trial {trial_dir} has different training argument hidden_size than the hyperparameter optimization search trials.')
+            if 'ffn_hidden_size' not in param_keys:
+                if getattr(hyperopt_args,'ffn_hidden_size') != trial_args['ffn_hidden_size']:
+                    raise ValueError(f'Manual trial {trial_dir} has different training argument ffn_hidden_size than the hyperparameter optimization search trials.')
+
+        for arg, space_parameter in matching_args:
+            if space_parameter not in param_keys:
```

Comment:
If the `linked_hidden_size` is in `param_keys`, should it be replaced with `hidden_size` and `ffn_hidden_size`? Because if `linked_hidden_size` is specified for hyperparameter optimization, this case we don't want the values of `hidden_size` and `ffn_hidden_size` be the same as the manual trail. However, setting `linked_hidden_size` as an element of `param_keys` cannot reflect this point. Please correct me if I misunderstand this part. Similar questions can be found in `init_lr` and `final_lr`.

Reply:
`param_keys` is the name of the parameter as labeled in the search space. `hyperopt_args` is the source of the model parameters in the new trials if they are not covered in the search space. `trial_args` is the model parameters as they appear in the manual trial. The goal of this block is to determine which parameters are controlled by the search space, and make sure for everything else, the parameter in `hyperopt_args` matches the one in `trial_args`.

`linked_hidden_size`, `hidden_size`, and `ffn_hidden_size` are handled in the code block L223-L233 instead of in the L235 loop. It's special behavior that get's handled there. If it's `linked_hidden_size`, then the hidden sizes can be any value as long as they're the same in the trial args. Then in L227-L233 it does the same check as the L235 loop for the regular hidden sizes, but only following the conditional that `linked_hidden_size` isn't in the search space.

`init_lr` and `final_lr` are still handled by the L235 loop. If you look at the tuple in L187 and L188, the label that's checked against the `param_keys` are the ratio versions.

Reply:
If `linked_hidden_size` is used, the `hidden_size` and `ffn_hidden_size` would be searched, but L235 will still want to make sure these two parameters in `hyperopt_args` are the same as in `trial_args`. However, these two parameters should have different values between `hyperopt_args` and `trial_args`.

Reply:
It wouldn't though because the hidden sizes aren't part of the L181 list. So L235 loop wouldn't cover them.

Reply:
You are right! Thanks for all the discussion. I have no further questions about this part. It makes sense to me.